src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs

```diff
@@ -0,0 +1,463 @@
+        /// <returns>Corresponding EdmPrimitiveType or null</returns>
+        public static IEdmPrimitiveType GetEdmPrimitiveTypeOrNull(Type clrType)
+        {
+            IEdmPrimitiveType primitiveType;
+            return _builtInTypesMapping.TryGetValue(clrType, out primitiveType) ? primitiveType : null;
+        }
+
+        /// <summary>
+        /// Builds Edm type reference to corresponding CLR type
+        /// </summary>
+        /// <param name="edmModel">Model to look up equivalent Edm type</param>
+        /// <param name="clrType">CLR type to look up</param>
+        /// <returns>IEdmTypeReference to the equivalent EdmType to the CLR type</returns>
+        public static IEdmTypeReference GetEdmTypeReference(this IEdmModel edmModel, Type clrType)
+        {
+            IEdmType edmType = edmModel.GetEdmType(clrType);
+            if (edmType != null)
+            {
+                bool isNullable = IsNullable(clrType);
+                return edmType.ToEdmTypeReference(isNullable);
+            }
+
+            return null;
+        }
+
+        // Mangle the invalid EDM literal Type.FullName (System.Collections.Generic.IEnumerable`1[[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]])
+        // to a valid EDM literal (the C# type name IEnumerable<int>).
+        public static string EdmName(this Type clrType)
+        {
+            // We cannot use just Type.Name here as it doesn't work for generic types.
+            return MangleClrTypeName(clrType);
+        }
+
+        /// <summary>
+        /// Returns the Edm formatted name of the CLR type
+        /// </summary>
+        /// <param name="clrType">CLR type</param>
+        /// <returns>Edm name of given CLR type</returns>
+        public static string EdmFullName(this Type clrType)
+        {
+            return String.Format(CultureInfo.InvariantCulture, "{0}.{1}", clrType.Namespace, clrType.EdmName());
+        }
+
+        public static bool IsNullable(Type type)
+        {
+            return !TypeHelper.IsValueType(type) || Nullable.GetUnderlyingType(type) != null;
+        }
+
+        internal static ClrTypeCache GetTypeMappingCache(this IEdmModel model)
+        {
+            ClrTypeCache typeMappingCache = model.GetAnnotationValue<ClrTypeCache>(model);
+            if (typeMappingCache == null)
+            {
+                typeMappingCache = new ClrTypeCache();
+                model.SetAnnotationValue(model, typeMappingCache);
+            }
+
+            return typeMappingCache;
+        }
+
+
+        /// <summary>
+        /// Finds a V4 type definition by the name of a V3 type definition
+        /// </summary>
+        /// <param name="model">V4 model to search</param>
+        /// <param name="t">V3 type</param>
+        /// <returns>Equivalent V4 Type or null if not found</returns>
+        public static IEdmType GetV4Definition(this IEdmModel model, Data.Edm.IEdmType t)
```

Comment:
>GetV4Definition [](start = 31, length = 15)

should we use "ODataVersion" to specify the version?

Reply:
We always want to get the OData4 definition, so there is no need to specify the version.